File: tests/combi/function_inline_simple.cpp
typedef struct {
  int x;
  int y;
  int z;
} vect;

vect vect_add(vect v1, vect v2) {
  return { v1.x + v2.x, v1.y + v2.y, v1.z + v2.z };
}

vect vect_mul(int d, vect v) {
  return { d * v.x, d * v.y, d * v.z };
}


int f(int x) {
  int a = x + x;
  return a + a;
}

int g(int x) {
  if (x > 0)
    return 1;
  else
    return 2;
}

int h(int x) {
  if (x > 0)
    return 1;
  return 2;
}

void m(int* p) {
  (*p)++;
}

int main() {
  int x = 3;
  int y = f(x);
  int z = g(x);
  int s = y + z;

  int u = h(x);
  int *q;
  m(q);

  vect a = {0,1,2};
  vect b = {3,4,5};
  vect c = vect_add (b, vect_mul(x, a));
  return 0;
}

vect vect_op(vect v) {
  if (true) {
    return v;
  }
  return v;
}

vect vect_op2(vect v) {
  vect res = {0,0};
  res.x = 1;
  return res;
}

void test_const_ret() {
  int x = 3;
  const int y = f(x);
  const int z = g(x);
  int s = y + z;
  const vect t = { 0, 1 };
  // const vect u = vect_mul(x, t);
  const vect v = vect_add(t, vect_mul(x, t));
  const vect w = vect_op(v);
  vect w2 = vect_op2(v);
  /*
    vect res = {0,0};
    res.x = 1;
    vect optitrust_r = res;
    vect w2 = optitrust_r;

    // do this step (possibly on the fly), unless const vect w2 and there are "goto exit"

    vect res = {0,0};
    res.x = 1;
    vect w2 = res;

    // do this step if previous step succeeded AND the "return" expr is a variable
    // AND (both w2 and res are const OR both are not const)

    vect w2 = {0,0};
    w2.x = 1;
  */
}

void test_const_arg() {
  const int x = 3;
  int y = f(x);
  int z = g(x);
  int s = y + z;
}

File: tests/combi/function_inline_simple.ml

open Optitrust
open Target

(* Fix this unit test *)
let _ = Run.doc_script_cpp (fun () ->

  !! Function.inline [cFun "f"];

)

"
int f(int x){ return x + 1;}

int main(){ 
  int x = 3;
  int y = f(x);
}
"
let _ = Run.script_cpp (fun _ ->

  (* !! Function.inline ~vars:(AddSuffix "${occ}") [occFirst;cFun "f?"]; *)
  !! Function.inline [nbMulti; cFun "vect_mul"];
  !! Function.inline [nbMulti; cFun "vect_add"];
  !! Function.inline  [nbMulti; cFun "vect_op"];
  (* TODO
   vect r;
  if (true) {
    {
      r = v;
      goto exit_body;
    }
  }
const vect w = r
*)

  !! Function.inline ~keep_res:true [nbMulti; cFun "vect_op2"];
  (* LATER: have an option to function inline for keeping "res" and not eliminating it ==> deactivate the final step
    vect res = {0,0};
    res.x = 1;
    vect w2 = res;

  *)

  (* inlining a function with single return *)
  !! Function.inline ~vars:(AddSuffix "${occ}") [nbMulti; cFun "f"];
  (* inlining a function with if else branches *)
  !! Function.inline [cTopFunDef "main"; cFun "g"];
  !! Function.inline ~resname:"r" [cTopFunDef "test_const_ret"; cFun "g"];

  (* inlining a function with one if branch *)
  !! Function.inline [nbMulti; cFun "h"];
  (* inlining a function of type void *)
  !! Function.inline [nbMulti; cFun "m"];

  (* with naming of the arguments *)
  !! Trace.alternative (fun () ->
    !! Function.inline  ~args:["v"] [nbMulti;cFunDef "main";cFun "f"];
    !!());

)
